### PR TITLE
Feature/7.x support

### PR DESCRIPTION
**Pull request by Mpdreamz**
**What issue does this PR address?**

- Adds 7.0 support (continuation of @kanadaj's PR #248 

- Adds `DetectElasticsearchVersion` to the sink that will detect the running cluster version. Something we now use to support sending `Esv6` templates to `Elasticsearch 7.x` and `Esv7` templates to `Elasticsearch 6.x` which should simplify upgrades. 

Adds an integration test project

- Spins up a 6.x and 7.x elasticsearch single node cluster in succession and asserts the default and the mixed mode through `DetectElasticsearchVersion` works.

**Does this PR introduce a breaking change?**

The base TFM is now `netstandard2.0` and `net461`.

**Please check if the PR fulfills these requirements**
- [x] The commit follows our [guidelines](https://github.com/serilog/serilog/blob/dev/CONTRIBUTING.md)
- [x] Unit Tests for the changes have been added (for bug fixes / features)

### CODE REVIEW

**Comment by Mpdreamz** [comment]
Note this PR deletes some `Durable` files that were duplicate by casing only, it does not remove the `Durable` feature.

File: src/Serilog.Sinks.Elasticsearch/Serilog.Sinks.Elasticsearch.csproj [hunk 1]
```diff
@@ -28,23 +28,6 @@
...
-    <PackageReference Include="Serilog.Sinks.File" Version="4.0.0" />
-    <PackageReference Include="Serilog.Sinks.PeriodicBatching" Version="2.1.1" />
-  </ItemGroup>
-
-  <ItemGroup Condition=" '$(TargetFramework)' == 'net45' ">
-    <Reference Include="System" />
-    <Reference Include="Microsoft.CSharp" />
-  </ItemGroup>
-
   <PropertyGroup Condition=" '$(TargetFramework)' == 'netstandard1.3' ">
```
Comment (mivano) [inline src/Serilog.Sinks.Elasticsearch/Serilog.Sinks.Elasticsearch.csproj]: Should this not be removed?

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs [hunk 2]
```diff
@@ -44,97 +43,205 @@ public class ElasticsearchTemplateProvider
                 case AutoRegisterTemplateVersion.ESv5:
                     return GetTemplateESv5(settings, templateMatchString);
```
Comment (mivano) [inline src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Does this imply supporting v5?

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs [hunk 3]
```diff
@@ -44,97 +43,205 @@ public class ElasticsearchTemplateProvider
...
                 case AutoRegisterTemplateVersion.ESv6:
-                    return GetTemplateESv6(settings, templateMatchString);
+                    return GetTemplateESv6(null, null, settings, templateMatchString);
+                case AutoRegisterTemplateVersion.ESv7:
+                    return GetTemplateESv7(null, null, settings, templateMatchString);
                 case AutoRegisterTemplateVersion.ESv2:
                     return GetTemplateESv2(settings, templateMatchString);
```
Comment (mivano) [inline src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Can also be removed?

File: src/Serilog.Sinks.Elasticsearch/Serilog.Sinks.Elasticsearch.csproj (hunk 1 above)
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Serilog.Sinks.Elasticsearch.csproj]: It should!

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs (hunk 2 above)
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Yes for now I left this in and move the deprecation of 5 and 2 into a separate PR as this one is quite big already

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs (hunk 3 above)
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Leaving this in for now, will follow up in a different PR

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs [hunk 4]
```diff
@@ -44,97 +43,205 @@ public class ElasticsearchTemplateProvider
...
+                                            {"MemberType", new {type = "integer"}},
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            };
+            mappings = discoveredVersion?.StartsWith("6.") ?? false ? new { _doc = mappings } : mappings;
```
[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Yes if `DetectElasticsearchVersion` is enabled. This allows you to send a v7 template to ES6 while upgrading or visa versa.

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticsearchSinkState.cs [hunk 5]
```diff
@@ -149,7 +170,11 @@ public void RegisterTemplateIfNeeded()
                     }
                 }
 
-                var result = _client.IndicesPutTemplateForAll<DynamicResponse>(_templateName, GetTempatePostData());
+                var result = _client.Indices.PutTemplateForAll<StringResponse>(_templateName, GetTempatePostData(),
```
[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticsearchSinkState.cs]: :+1:

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs [hunk 6]
```diff
@@ -44,97 +43,205 @@ public class ElasticsearchTemplateProvider
...
+                                    "ExceptionMessage", new
                                     {
-                                        match = "*",
-                                        match_mapping_type = "string",
-                                        mapping = new
+                                        type = "object",
+                                        properties = new Dictionary<string, object>
                                         {
-                                            type = "text",
+                                            {"MemberType", new {type = "integer"}},
```
[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Let me double check what happened with the mappings.

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs (hunk 2 above)
Comment (mivano) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Sounds like a good plan

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs [hunk 7]
```diff
@@ -44,97 +43,205 @@ public class ElasticsearchTemplateProvider
...
-                                        path_match = @"fields\.[\d+]$",
-                                        match_pattern = "regex",
-                                        mapping = new
+                                        raw = new
                                         {
-                                            type = "text",
+                                            type = "keyword",
                                             index = true,
-                                            norms = false
+                                            ignore_above = 256
```
[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: The diff here is a bit confusing as the lines it matches against are for the v5 template but the template should remain unchanged for 7 from 6 apart from removing `__default__` nested type block.

File: src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs]: Its what its currently mapped to: https://github.com/serilog/serilog-sinks-elasticsearch/blob/dev/src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticSearchTemplateProvider.cs#L135

**Comment by Mpdreamz** [comment]
Updated the PR @mivano thanks for the review.

I am on holiday for 2 weeks so won't be able to merge this in if all good so please feel free to do so!

**Comment by mivano** [comment]
Thanks @Mpdreamz and have a good holiday. I will merge it to dev so we have some test packages.

**Comment by DanVicarel** [comment]
@mivano @Mpdreamz Any idea when a stable v7.2 of this NuGet will be deployed with these changes?

**Comment by mivano** [comment]
I pushed an alpha package to nuget so we can gather some feedback: https://www.nuget.org/packages/Serilog.Sinks.ElasticSearch/8.0.0-alpha0025
I rather not merge it to master just yet, certainly not with the holidays upcoming.

**Comment by kanadaj** [comment]
@mivano wouldn't it be better to try to match versions of the Sink to versions of the SDK and/or Elasticsearch itself? Just wondering.

**Comment by mivano** [comment]
It is breaking enough that it warrants a major version bump. The 7 or 8 does not also match the ES product version. It still has support for 6 for example.

**Comment by mivano** [comment]
Version 8 is now published: https://www.nuget.org/packages/Serilog.Sinks.ElasticSearch/8.0.0

**Comment by hypertrends** [comment]
Good Afternoon, I am trying to connect to Elastic Search using the latest 8.0.0 and I don't see any log entries in my Elastic Search. Is there a different configuration I have to do? Here's what my code looks like for Service Fabric w/ Serilog:

`Log.Logger = new LoggerConfiguration()
                            .WriteTo.ApplicationInsights(appInsightsKey, TelemetryConverter.Events)
                             .WriteTo.Elasticsearch(
                                new ElasticsearchSinkOptions(
                                    new Uri(elasticUrl))
                                {
                                    AutoRegisterTemplate = true,
                                    AutoRegisterTemplateVersion = AutoRegisterTemplateVersion.ESv7,
                                    //BufferBaseFilename = "./logs/buffer",
                                    IndexFormat = index,
                                    FailureCallback = e =>
                                    {
                                        Console.WriteLine("Unable to submit event " + e.Exception);
                                    },
                                    EmitEventFailure = EmitEventFailureHandling.WriteToSelfLog |
                                                       EmitEventFailureHandling.WriteToFailureSink |
                                                       EmitEventFailureHandling.RaiseCallback,
                                    ModifyConnectionSettings = x =>
                                        x.BasicAuthentication(elasticUsername, elasticPassword)
                                })
                            .CreateLogger();`

The index is of a simple 'test-{0:yyyy.MM}' format. I have commented out the BufferBaseFileName in the snippet above but uncommenting it doesn't make any changes.

What's interesting is that I have an ASP.NET Core project that does this just fine so I know it isn't a v7 issue. I read somewhere in an issues list that only public facing IP systems are able to make logs? Please also note that there are 5 instances of the same service in Service Fabric. Could that be affecting things?

Would love to get some feedback. I've spent quite a bit of time trying to figure this out.

**Comment by staticdev** [comment]
@hypertrends same here.

**Comment by hypertrends** [comment]
Hello @staticdev - let me check that. As you can see in my snippets, I definitely have that information set right. I will review my log levels to see what's up.

**Comment by SamilMehdiyev** [comment]
@staticdev , @hypertrends

I had the same problem. After few hours investigation and debugging, I founded that Assembly Version of Serilog.Sinks.Elasticsearch was old. I am using 8.0.0 version, but in file displayed 6.0.0. 

To fix this problem, I did the followings:

1) Clone the repository
2) Open it in MS Visual Studio
3) Changed version from 6.0.0 to 8.0.0 in AssemblyInfo.cs file
4) Build the project
5) Use new dll in my own project instead of package which downloaded from NuGet

Hi @mivano , I would like to push my change in dev brach. Or Kindly ask you to update assembly version and publish new version, so we can use it from NuGet.

Thanks, hope it will help.

**Comment by hypertrends** [comment]
Good Morning @SamilMehdiyev - wow, I hadn't even thought about that. Looking at it right now :) Thank you so much!

**Pull request merged**